spec/Classes.tex

```diff
@@ -784,12 +787,54 @@ \subsection{The Method Receiver and the {\em this} Argument}
 is \chpl{c1}.
 \end{chapelexample}
 
-The optional \sntx{this-intent} declares the intent by which the receiver
-argument will be passed.  When the value of \sntx{this-intent} is \chpl{param},
-it implies that the function can only be applied to param objects of the given
-type binding.  If the value is \chpl{ref}, the argument will be passed by
-reference---allowing modification of \chpl{this}.  Otherwise, the argument will
-be passed with the blank intent as specified in \rsec{The_Blank_Intent}.
+The optional \sntx{this-intent} declares how the receiver argument is
+constrained or will be passed for a given method call.
+
```

Comment:
I feel OK in agreeing with your summary for Smalltalk, Self, C++, Java, Chapel.


CLOS is really in it’s own camp in all of this and I may have err’d in even mentioning
it in this conversation.



On the one hand CLOS shares the Smalltalk/SELF notion of classes being “simply”
instances of some type that derives from Class with a particular responsibility to create
other objects.  This is what provoked me to mention CLOS.

On the other hand CLOS does not associate methods with classes.  Instead it has
generic-functions that are specialized by the types of the formals.  I’m a bit rusty on
CLOS, esp. as I was primarily a lisp user in the ZetaLisp days, but I think it’s appropriate
to claim that one can write a generic-function with one method that is specialized on the
Class type and a separate method that is specialized on an instance of the  type generated
by the class.   But I suspect this isn’t really the way a CLOS programmer would be thinking
about things.


But finally CLOS does have a model for thinking about “instance slots” vs “class slots”
where slot means field.  But this is wrapped up with how classes are defined and doesn’t
have anything to do with CLOS’s sense of methods.


Mike



On Jul 7, 2015, at 5:31 PM, Brad Chamberlain <notifications@github.com<mailto:notifications@github.com>> wrote:


In spec/Classes.tex<https://github.com/chapel-lang/chapel/pull/2144#discussion_r34106648>:

> @@ -784,12 +787,54 @@ \subsection{The Method Receiver and the {\em this} Argument}
>  is \chpl{c1}.
>  \end{chapelexample}
>
> -The optional \sntx{this-intent} declares the intent by which the receiver
> -argument will be passed.  When the value of \sntx{this-intent} is \chpl{param},
> -it implies that the function can only be applied to param objects of the given
> -type binding.  If the value is \chpl{ref}, the argument will be passed by
> -reference---allowing modification of \chpl{this}.  Otherwise, the argument will
> -be passed with the blank intent as specified in \rsec{The_Blank_Intent}.
> +The optional \sntx{this-intent} declares how the receiver argument is
> +constrained or will be passed for a given method call.
> +


To make sure I'm understanding correctly, I think you're saying:

  *   In Smalltalk, Self, and CLOS, if a method is defined on a class (rather than an instance of that class), it can only be called on the class, not on instances of that class.
  *   The currently-implemented behavior of Chapel is therefore closer to Smalltalk, Self, and CLOS
  *   However, in most other contexts, Chapel is more like C++/Java than Smalltalk/Self/CLOS. Therefore we may want to choose not to differ from them in this regard (suggesting the implementation would need to change).

Did I get that right?

—
Reply to this email directly or view it on GitHub<https://github.com/chapel-lang/chapel/pull/2144/files#r34106648>.